### PR TITLE
[PVR] PVR database: Fix error "Abort due to constraint violation. Query: INSERT INTO channels..."

### PR DESCRIPTION
**Pull request by ksooo**
This error has been reported by several users for a long time and I think I now have found the cause and was finally able to fix it.

The problem is quite simple: `CPVRDatabase::Persist(CPVRChannel &channel)` used `channel.ChannelID()` to decide whether a channel was already contained in the `channels`table. But the provided channel id sometimes was not correct. For the given bug scenario it is `-1` which means it is a new channel, but there actually are data for this channel in the `channels` table. The following `INSERT into channels` obviously will fail (when comitting the queued queries later on). How can this happen? Well, the channel id is picked from a different table earlier during pvr init - `map_channelgroup_channels` and it seems that `channels` and `map_channelgroup_channels` can get out of sync - maybe when Kodi crashes while deleting channels from the database, right after data from `map_channelgroup_channels` are deleted and before data from `channels` can be deleted.

I'm not SQL expert and maybe the root cause of the data getting out of sync can be prevented? I'm open for suggestions. My PR just prevents the error in case db corruption already happened, by checking for the channel in the `channels`table instead of trusting the provided channel id.

@Jalle19 mind taking a look at the code? You were trying to fix this years ago with no success, IIRC.
@chewitt you said you could runtime test this?
@MilhouseVH can you include this PR in your next build, so that @chewitt has something to test?

### CODE REVIEW

File: xbmc/pvr/PVRDatabase.cpp [hunk 1]
```diff
@@ -788,10 +789,6 @@ bool CPVRDatabase::Persist(CPVRChannel &channel, bool bCommit)
 
   if (QueueInsertQuery(strQuery))
   {
-    /* update the channel ID for new channels */
-    if (channel.ChannelID() <= 0)
-      channel.SetChannelID(static_cast<int>(m_pDS->lastinsertid()));
-
```
Comment (Jalle19) [inline xbmc/pvr/PVRDatabase.cpp]: Why is this not needed anymore?

File: xbmc/pvr/PVRDatabase.cpp [hunk 2]
```diff
@@ -757,10 +757,11 @@ bool CPVRDatabase::Persist(CPVRChannel &channel, bool bCommit)
     return bReturn;
   }
 
-  std::string strQuery;
   CSingleLock lock(m_critSection);
 
-  if (channel.ChannelID() <= 0)
+  std::string strQuery = PrepareSQL("iUniqueId = %u AND iClientId = %u", channel.UniqueID(), channel.ClientID());
```
Comment (Jalle19) [inline xbmc/pvr/PVRDatabase.cpp]: Maybe add a comment here so someone doesn't refactor this later in search for performance gains?

**Comment by chewitt** [comment]
@ksoo no issues (or SQL errors) seen with this so far :)

File: xbmc/pvr/PVRDatabase.cpp (hunk 1 above)
Comment (ksooo) [reply xbmc/pvr/PVRDatabase.cpp]: I knew you would ask:

1. As we 're not insertung instantly, this does not work (never did) as `m_pDS->lastinsertid()` gets not increased.
2. SetChannelID is called later on, after all INSERTs are done. I double checked this both in debugger and db browser. => https://github.com/xbmc/xbmc/blob/master/xbmc/pvr/PVRDatabase.cpp#L659

File: xbmc/pvr/PVRDatabase.cpp (hunk 2 above)
Comment (ksooo) [reply xbmc/pvr/PVRDatabase.cpp]: Good point, will add a comment.

**Comment by ksooo** [comment]
> no issues (or SQL errors) seen with this so far :)

 cool. thanks for the info.

File: xbmc/pvr/PVRDatabase.cpp (hunk 1 above)
Comment (Jalle19) [reply xbmc/pvr/PVRDatabase.cpp]: Cool

**Review by Jalle19 — APPROVED** [review]

**Comment by tamland** [comment]
> I'm not SQL expert and maybe the root cause of the data getting out of sync can be prevented? I'm open for suggestions.

Simple: use transactions when deleting

**Comment by ksooo** [comment]
>Simple: use transactions when deleting

 Simple in theory, yes. Not simple, as I do not know how to code that.

**Comment by tamland** [comment]
It's used in all the other DBs (if you need examples). Just call the BeginTransaction member, execute all your deletes/updates/whatever, and finally CommitTransaction. One should really, *always*, do this when executing more than one query..

**Comment by ksooo** [comment]
@tamland Thanks. Will do this in a followup PR. I guess there will be "some more" problems wrt. transactions in PVR database handling.

**Comment by ksooo** [comment]
> One should really, always, do this when executing more than one query..

 Agreed, but please don't shoot the messenger. This has all been written long before I joined the project. :-)

**Pull request merged**